crates/noirc_frontend/src/monomorphization/mod.rs

```diff
@@ -745,15 +767,65 @@ impl<'interner> Monomorphizer<'interner> {
         id: node_interner::ExprId,
     ) -> ast::Expression {
         let func = Box::new(self.expr(call.func));
-        let arguments = vecmap(&call.arguments, |id| self.expr(*id));
+        let mut arguments = vecmap(&call.arguments, |id| self.expr(*id));
+        let hir_arguments = vecmap(&call.arguments, |id| self.interner.expression(id));
         let return_type = self.interner.id_type(id);
-        let return_type = Self::convert_type(&return_type);
+        let return_type = self.convert_type(&return_type);
         let location = call.location;
 
+        if let ast::Expression::Ident(ident) = func.as_ref() {
+            if let Definition::Oracle(name) = &ident.definition {
+                match name.as_str() {
+                    "println" => {
+                        self.append_abi_arg(&hir_arguments[0], &mut arguments, false);
```

Comment:
Could you expand on this a bit? It looks like we're relying on some weird serialization here with serializing the type of the argument and pushing that as a string. It seems somewhat roundabout.

Matching on the literal function name here will also fail when higher order functions are used:
```rs
let p = dep::std::println;
p(42);
```
Would this cause a compiler crash?

If we could remove the function name check and rely on the type system somehow instead we could get rid of the serialization as well.

Reply:
The snippet you posted does not crash and printed "0x2a" as expected. 

This was done due to the need to attach some sort of type metadata that tells nargo how to print what is passed to a println oracle. A foreign call in Brillig is simply a list of field elements so we need some way to de/serialize the type metadata from Brillig foreign calls. We already have an AbiType as part of our `noirc_abi` that handles de/serialization of a list of field elements with associated high level types to/from JSON. So I decided to follow Noir's abi serialization to enable println to support many types without having to rely on a new serialization strategy for println. 

I will look at relying on the type system rather than the function name check, but I don't think there is anyway to get around the serialization as the caller of the foreign function needs some way to interpret the list of field elements that has been returned from a paused VM instance. I'll make sure to document this method and the purpose of the serialization as it is not clear just looking at the method alone.

Reply:
I left a comment explaining the purpose of this method